### PR TITLE
Doc: Add iterator reference

### PR DESCRIPTION
**Pull request by notch1p**
addresses #241 #251

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -461,6 +461,75 @@ fn init {
...
+Almost all `Iter` implementations are identical to that of `Bytes`, the only
+main difference being the code block that actually does the iteration.
+
+Iterator provides a unified way to iterate through data structures, and they
+can be constructed at basically no cost: as long as `fn(yield)` doesn't
+execute, the iteration process doesn't start.
+
+Underneath the abstractions of `map` and such, is a `self.run()`. Whenever a
+`self.run()` is called, the iteration starts. Chaining such methods might be
+visually pleasing, but do notice the heavy work deep down.
```
Comment (peter-jerry-ye) [inline README.md:525]: There's no such API called `run` and `map` itself will not trigger an iteration: it will only construct an `Iter`.

Maybe rephrase this part and mention the concept of 'consuming' an `Iter`.

File: README.md [hunk 2]
```diff
@@ -461,6 +461,75 @@ fn init {
...
+An iterator is an object that traverse through a sequence while providing access
+to its elements. Traditional OO languages like Java's `Iterator<T>` use `next()`
+`hasNext()` to step through the iteration process, whereas functional languages
+(JavaScript's `forEach`, Lisp's `mapcar`) provides a high-order function which
+takes an operation and a sequence then consumes the sequence with that operation
+being applied to the sequence. The former is called _external iterator_ (visible to user) and the
+latter is called _internal iterator_ (invisible to user).
+
+The built-in type `Iter[T]` provides internal iterator support for
+MoonBit. The type `Iter[T]` is basically a type alias for `((T) -> IterResult)-> IterResult`,
```
Comment (peter-jerry-ye) [inline README.md:475]: You might need to explain the concept of `IterResult`. Check with others to see how `Iter` will support `Error`

File: tour.md [hunk 3]
```diff
@@ -33,11 +33,26 @@ This automatically installs MoonBit in `$HOME/.moon` and adds it to your `PATH`.
...
+status bar that configures the target.
+
+![compile-target](./imgs/compile-targets.png)
+
+While our main focus being the wasm-gc, debugging is only supported by
+javascript backend for now. Be sure to try out the interactive debugging feature
+provided by our extension.
+
+### Project Structure
+
```
Comment (peter-jerry-ye) [inline tour.md:50]: Better to put it in another PR: one modification per PR.

**Comment by notch1p** [comment]
all problems should be fixed now.

File: README.md [hunk 4]
```diff
@@ -461,6 +461,103 @@ fn init {
...
+- `filter`: _lazy_ Filters the elements of the iterator based on a predicate function.
+- `map`: _lazy_ Transforms the elements of the iterator using a mapping function.
+- `concat`: _lazy_ Combines two iterators into one by appending the elements of the second iterator to the first.
+
+Methods like `filter` `map` are very common on a sequence object e.g. Array.
+But what makes `Iter` special is that any method that constructs a new `Iter` is
+_lazy_, i.e. calling such methods does not start the iteration process as it's
+wrapped inside a function internally, thus basically no cost. That's what makes `Iter` superior for
+traversing through sequence. MoonBit encourages user to pass an `Iter` across
+functions instead of the sequence object itself.
```
Comment (Guest0x0) [inline README.md:504]: The benefit of `Iter` is that it requires no allocation for intermediate result. Laziness is a consequence of no-intermediate-allocation, but laziness itself does not imply no intermediate allocation. So it's a bit imprecise here to credit laziness for the efficiency of `Iter`

Comment (notch1p) [reply README.md]: good call. I've clarified this.
